elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,405 @@
+/**
+ * // TODO: Big tree comment.
+ */
+
+(function() {
+  /**
+   * Define keycodes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    UP: 38,
+    DOWN: 40,
+    LEFT: 37,
+    RIGHT: 39,
+    HOME: 36,
+    END: 35,
+    SPACE: 32,
+    ENTER: 13,
+  };
+
+  /**
+   * A helper to quickly identify `<dash-treeitem>` nodes
+   */
+  function isTreeItem(node) {
+    return node.nodeName.toLowerCase() === 'dash-treeitem';
+  }
+
+  /**
+   * A helper to quickly identify ARIA states.
+   * Note that ARIA attributes require a literal string value of "true"/"false".
+   */
+  function isExpanded(node) {
+    return node.getAttribute('aria-expanded') === 'true';
+  }
+
+  // `dashTreeItemCounter` counts the number of `<dash-treeitem>` instances
+  // created. The number is used to generated new, unique IDs.
+  let dashTreeItemCounter = 0;
+
+  class DashTreeItem extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'treeitem');
+
+      // If the element doesn't already have an ID, generate one for it.
+      // This will make it easier to set the element as active using
+      // aria-activedescendant.
+      if (!this.id)
+        this.id = `dash-treeitem-generated-${dashTreeItemCounter++}`;
+
+      // If the element contains a child then it's a parent node
+      if (this.firstElementChild) {
+          // If the element is not explicitly already expanded by the user, then
+          // set it to closed.
+          if (!isExpanded(this))
```

Comment:
I see. Previously I had a few of these as class methods but because I use them in the recursive _findTreeItems function it got a little wonky having to keep rebinding to `this`. Let me see if it's a little simpler now that I only have this one method.